Address code duplication issues by making object leaking unconditional

sb98052 opened the pull request:
Release notes: Fixed an issue that was causing code duplication

This change is modeled after #2125 and makes object leaking unconditional in the same ways as declarative bindings now are. It resolves #2083 and its conditional variant (checked in as the test ConditionallyLeakedObject.js). Additionally, it renames the _isHavoced flag and the API to it to use the term "leak" instead of "havoc" to clarify the difference between tainting the future analysis of an object on the one hand, and losing information about its currently tracked values on the other.

Possible follow ups:
- Selective object serialization via literals even in the face of leaking
- More complete test coverage over getters and setters

NTillmann commented on src/values/ObjectValue.js:
```diff
@@ -405,15 +405,15 @@ export default class ObjectValue extends ConcreteValue {
-  mightBeHavocedObject(): boolean {
-    return this._isHavoced.mightBeTrue();
+  isLeakedObject(): boolean {
+    return this._isLeaked.equals(this.$Realm.intrinsics.true);
```
```
let isLeaked = this._isLeaked; // cache property access
invariant(isLeaked instanceof BooleanValue); // to make sure that joining always does the right thing
return isLeaked.value;
```

And something similar for `isNotLeakedObject`.

NTillmann commented on src/values/ObjectValue.js:
```diff
@@ -405,15 +405,15 @@ export default class ObjectValue extends ConcreteValue {
 
-  mightNotBeHavocedObject(): boolean {
-    return this._isHavoced.mightNotBeTrue();
+  isNotLeakedObject(): boolean {
```
Or you could cut `isNotLeakedObject` altogether, as it's really just `!isLeakedObject` now.

NTillmann commented on src/methods/join.js:
```diff
@@ -806,7 +806,40 @@ export class JoinImplementation {
+
     let join = (b: PropertyBinding, d1: void | Descriptor, d2: void | Descriptor) => {
+      if (b.object instanceof ObjectValue && b.key === "_isLeaked") {
+        invariant(ObjectValue.trackedPropertyNames.includes("_isLeaked"));
```
This checks a static property doing a linear search every time this happens. Do you really want to keep this invariant?

sb98052 replied:
The point was to verify that the tag used is a keyword, so that if somebody renames it or tracks. But I can do something to make sure it's only called once.

NTillmann commented on src/methods/join.js:
```diff
@@ -806,7 +806,40 @@ export class JoinImplementation {
+    };
+
     let join = (b: PropertyBinding, d1: void | Descriptor, d2: void | Descriptor) => {
+      if (b.object instanceof ObjectValue && b.key === "_isLeaked") {
```
Hm, so in `PropertyBinding`, the `object` is defined to be `AbstractObjectValue`, and thus you need to do the cast here. Okay. But I wonder if that's really the most appropriate type definition.
For fun, I removed the `| AbstractObjectValue`, and ran flow again --- no violation!
Now, of course that doesn't have to mean much, but I wonder if we can make this new code cleaner by if can simply "fix" the type definition, after manually reviewing all assignments?

sb98052 replied:
See #2208

NTillmann commented on src/methods/join.js:
```diff
@@ -806,7 +806,40 @@ export class JoinImplementation {
+        return d1;
+      } else {
+        // Make the leak unconditional
+        let isTrue = bv => bv instanceof BooleanValue && bv.equals(realm.intrinsics.true);
```
just `bc.value` instead of the `.equals` call

hermanventer replied:
Every value has already has method mightNotBeTrue, also isTrue(bv) => !bv.mightNotBeTrue() but the other way around, so using the latter is probably better.

NTillmann commented on test/serializer/optimized-functions/ConditionallyLeakedBinding.js:
```diff
@@ -0,0 +1,17 @@
+(function () {
+    function f(c, g) {
+        let x = 23;
```
There is no object here, what does this test?

sb98052 replied:
I tried a slew of different tests to break your unconditional leak declarative bindings patch. This one I decided to keep (don't quite remember why). If you think it's covered by something else, I can drop it.

NTillmann commented on test/serializer/optimized-functions/ConditionallyLeakedObject.js:
```diff
@@ -0,0 +1,17 @@
+  }
+
+  global.__optimize && __optimize(f);
+  inspect = function() { return f(true, h => h()); }
```
Just returning an object from the inspect function will not cause a deep inspection. You want to create a string which contains `.now` and `.foo`.

hermanventer replied:
f expects its first argument to be a function.

NTillmann commented on test/serializer/optimized-functions/LeakObjectWithSetter.js:
```diff
@@ -0,0 +1,11 @@
+  }
+
+  global.__optimize && __optimize(f);
+  inspect = () => { f( h => h ()); }
```
Make inspect function do a deep inspection.

NTillmann commented on test/serializer/optimized-functions/LeakedObjectCodeDuplication.js:
```diff
@@ -0,0 +1,9 @@
+}
+
+global.__optimize && __optimize(f);
+inspect = function() { return f(true, h => h()); }
```
Same here.

sb98052 replied:
Will do.

hermanventer commented on test/serializer/optimized-functions/LeakedObjectCodeDuplication.js:
```diff
@@ -0,0 +1,9 @@
+}
+
+global.__optimize && __optimize(f);
+inspect = function() { return f(true, h => h()); }
```
f only takes one argument.

NTillmann commented on src/intrinsics/ecma262/Object.js:
```diff
@@ -59,7 +59,7 @@ function handleObjectAssignSnapshot(
     }
 
-    if (frm instanceof ObjectValue && frm.mightBeHavocedObject()) {
+    if (frm instanceof ObjectValue && frm.isLeakedObject()) {
```
A big motivation behind Sapan's changes is to make leaking unconditional, so an object either has leaked, or it didn't leak.

hermanventer replied:
I don't get this. Why?

sb98052 replied:
It is more appropriate if we retain flow sensitivity for the _isLeaked flag, it is not more appropriate if we drop this flow sensitivity. 

The purpose of this pull request is to deal with competition between materialization of object properties via the generator timeline, and interpretation, which is independent of the timeline. The outcome is duplicated code caused by the creation of object properties both via materialization and object initialization. For the non-branched case, this is easy to fix - if an object is leaked, you suppress the serialization of the second kind of initialization. For the branched case, it is more involved (and may also affect correctness - although I could not come up with a test case equivalent to #2007). Nikolai's idea was to handle it by eliminating flow-sensitive "materialization." If an object is leaked, it is leaked and materialized along all consequent as well as alternate paths.

This effectively eliminates flow-sensitive leaking in practice, so if we do this, then there is no point in tracking it via abstract conditionals.

sb98052 commented:
I've addressed the issues brought up in the comments, and also significantly refactored the leak-accounting logic. Instead of dealing with leaks where modified properties are joined (joinModifiedProperties), leaked objects are now collected and passed on for joining at a more opportune point, when the rest of the effects have been applied. This is necessary to carry out materialization correctly and in the right order in the face of other accesses to the same object. Leaked objects have been made an optional field of the Effects structure, and the materialization part of leaking and havocing has been moved to ObjectValue.js.

hermanventer commented on src/methods/join.js:
```diff
@@ -807,18 +831,70 @@ export class JoinImplementation {
+      b: PropertyBinding,
+      d1: void | Descriptor,
+      d2: void | Descriptor
+    ): [Descriptor | void, IntermediateObjectValue | void] => {
```
Given our existing naming scheme, it is really confusing for an (ObjectValue, Generator) pair to be referred to as an IntermediateObjectValue. Every other type that ends with the Value suffix is a subclass of Value. This one is not. That is surprising and a trap for the unwary.

hermanventer commented on src/methods/join.js:
```diff
@@ -807,18 +831,70 @@ export class JoinImplementation {
+      } else {
+        // Make the leak unconditional
+        let isTrue = bv => bv instanceof BooleanValue && bv.value;
+        let _d1Value = isTrue(d1.value);
```
We already have !d1.value.mightNotBeTrue(). Also, the _ prefix is used for private object properties and using it for a local name seems strange. d1ValueIsTrue would be a better name.

hermanventer commented on src/methods/properties.js:
```diff
@@ -107,7 +107,7 @@ function InternalUpdatedProperty(realm: Realm, O: ObjectValue, P: PropertyKeyVal
   if (P instanceof SymbolValue) return;
   if (P instanceof StringValue) P = P.value;
-  invariant(!O.mightBeHavocedObject()); // havoced objects are never updated
+  invariant(!O.isLeakedObject()); // havoced objects are never updated
```
leaked objects are

hermanventer commented on src/realm.js:
```diff
@@ -93,13 +94,15 @@ export class Effects {
     propertyBindings: PropertyBindings,
-    createdObjects: CreatedObjects
+    createdObjects: CreatedObjects,
+    leakedObjects?: IntermediateObjects
```
Why not call the type "LeakedObjects"?

hermanventer commented on src/utils/havoc.js:
```diff
@@ -173,57 +166,26 @@ class ObjectValueHavocingVisitor {
-              }
-            }
-          }
+    if (!obj.isLeakedObject()) {
```
I'm uncomfortable with this. Havocing an object is not the same as leaking it. This conflates the two concepts.

sb98052 replied:
This comment provides no information, beyond your being uncomfortable with this. The point of the renames I did was to clarify the difference in the code between these two concepts. Do you disagree with the changed use at this code site? Do you disagree with it in general? Please be more precise.

hermanventer replied:
A routine that is named "havoc" ends up "leaking" an object. That smells wrong.

NTillmann replied:
This entire file / class should get renamed away from "havoc" to "leaking". Materialization and havocing are side effects of leaking.

hermanventer replied:
If we come across a use case where we need to havoc something without leaking it, this approach will bite us. Right now, I'm not convinced that there is such a use case, so this may turn out to be OK. It just smells a bit off to me.

hermanventer commented on src/values/ObjectValue.js:
```diff
@@ -404,16 +408,70 @@ export default class ObjectValue extends ConcreteValue {
-  mightBeHavocedObject(): boolean {
-    return this._isHavoced.mightBeTrue();
+  /* Emits generator entries to set the values of properties in an object */
+  materializeInGenerator(generator: Generator, getHeapInspector?: void => HeapInspector, emitDeletes?: boolean): void {
```
This seems to do the same kind of thing as ObjectValue.getSnapSnot. That seems like a great pity to me.

sb98052 replied:
The great pity to me is that ObjectValue.getSnapshot is not documented, as is not the model of which it is a part. Nonetheless, I will see if I can somehow decipher its behavior and reconcile it with my changes.

trueadm replied:
@sb98052 I can fill you in on how getSnapshot works if that helps? Feel free to ping me and we can setup a VC or chat over workplace. Otherwise, I'll be sure to put up a PR tomorrow adding some comments around it if that helps.

sb98052 replied:
Thanks. Comments would certainly be helpful. Right now, the `materializeInGenerator` function factorizes out code that was already in the (old-terminology-) havocing implementation. As such, if there is any redundancy between these two things, then it is pre-existing, and out of scope for this pull request. I would be happy to take on the task of reconciling the two things in a separate task.

sb98052 commented on src/methods/join.js:
```diff
@@ -807,18 +831,70 @@ export class JoinImplementation {
+    // Joins two property bindings of type "_isLeaked". If an object is leaked
+    // on one side, it is also leaked on the other side. Returns [x, y] where
+    // x is the joined descriptor, and y is set to the IntermediateObjectValue consisting of generator and object if the object leaked.
+    let joinLeakedObjectPropertyAndAdvise = (
```
The needless emphasis on your experience of being befuddled and the characterization thereafter detracts from what is a simple question, "Why the term Advice?" It was a loose reference to this - https://en.wikipedia.org/wiki/Advice_(programming). What is being proffered is some logic or computation that is to be woven in at a future join point.

hermanventer replied:
I apologize if I have offended you. If you do not welcome the unguarded impressions of someone reading your code, I'll refrain. No offense was intended. It is just me being my authentic self.

sb98052 replied:
No offense was taken. The unguardedness of my own impressions is just as amenable to the analysis of code as it is to the code itself. If this is unacceptable, then I shall refrain. I hope that I correctly interpreted your question and was able to clarify my terminology. In the next version (to be put up soon) I have changed the name of this function to `computeJoinedLeakPropertyAndLeakedObject`.

sb98052 commented:
Take 3. I have addressed the review comments. The main other change is the addition of the materialization path to `singletons.js` to break the flow dependency cycle between `havoc.js` and `realm.js`. I spent a fair bit of time trying various gymnastics, such as declaring the HeapInspector class, breaking it up, and moving code around between `ObjectValue.js`, `join.js` and `havoc.js`, but could not find a way short of this approach to break the dependency.

This changeset raises a number of future tasks, but I feel that they should be handled separately in the interest of keeping this set of changes focused on the issues it is intended to address. Some of these are:

- Factoring out redundant behavior between getSnapshot and materialization
- Making a deep pass through the code base and changing the terms havoc, leak and materialize to the appropriate term

I would be happy to add issues for these items and assign them to myself.

hermanventer commented on src/methods/join.js:
```diff
@@ -709,25 +717,26 @@ export class JoinImplementation {
       let [aVal3, bVal] = fold(key, aVal1, aVal2);
       acc.map.set(key, aVal3);
-      if (acc.set !== undefined && bVal !== undefined) acc.set.add(bVal);
+      if (bVal !== undefined) acc.set.add(bVal);
```
The previous version of this made sense in as much as that it avoided building a set unless it was actually wanted. Do you now make use of the set for every call of foldMaps?

sb98052 replied:
I teetered between the two options, both of which are unsatisfying. With one you get a redundant Set, with the other you spill the void type outwards, all the way to the LeakedObjects type, get asymmetry in the Set, Map tuple (in the MapAndSet data structure, one can be undefined but the other can't). That invariant was the reason for the redundant Set in the previous version. I dislike both options, and chances are, every time somebody points out the infelicity of one, I will rush towards the other.

hermanventer replied:
Perhaps this calls for two separate methods?

sb98052 replied:
I think I have found a good middle ground, and fixed a typing bug that helps eliminate some unnecessary annotations and invariants.

hermanventer commented on src/types.js:
```diff
@@ -163,6 +171,8 @@ export type PropertyBinding = {
   internalSlot?: boolean,
 };
 
+export type LeakedObject = {| object: ObjectValue, generator: Generator |};
```
A comment explaining the role of the generator will be very helpful to a new reader.

hermanventer commented on src/realm.js:
```diff
@@ -1485,9 +1486,9 @@ export class Realm {
-        invariant(ioVal !== undefined);
-        ioVal.object.materializeInGenerator(ioVal.generator);
+      for (let lo of leakedObjects) {
+        invariant(lo !== undefined);
```
Best practice for invariants is to explain why you believe they can never be false. In this instance, I cannot verify this unless I leave the diff view (of this commit) and start reading the overall code. A comment would have saved me a bunch of time.

As it is, I can't quite figure out how lo can ever become undefined. So my question is: Why is LeakedObjects defined to allow undefined elements? In particular, what is the use case and which test case exercises that use case?

trueadm replied:
I believe this was just a rename from `ioVal` which was the same invariant as before. So maybe @NTillmann can explain more why `leakedObjects` is a union with `void`. We should probably refine that type so that there is no union with `void`?

sb98052 commented:
I've addressed the last batch of review comments.

NTillmann commented on src/serializer/ResidualHeapSerializer.js:
```diff
@@ -1674,13 +1674,25 @@ export class ResidualHeapSerializer {
+    let isLeaked = val.isLeakedObject();
+
+    let shouldDropAsAssignedProp = (descriptor: Descriptor) =>
+      isLeaked && !(isFunction(descriptor.get) || isFunction(descriptor.set));
```
This special case for accessor properties here (and the corresponding code in the visitor) only exists because of this comment `// TODO: Deal with accessor properties` in the materialization code, right? 

If so, please create an issue, and have comments like `TODO #Issue` in all the relevant places in the code.

sb98052 replied:
Done. Yes, things get complicated with getters and setters, especially if they cross reference other properties in the object.

NTillmann commented on src/realm.js:
```diff
@@ -1485,6 +1490,13 @@ export class Realm {
+    // materialize leaked objects
+    if (leakedObjects !== undefined && leakedObjects.size !== 0) {
+      for (let lo of leakedObjects) {
+        Materialize.materialize(this, lo);
```
Is this the main place where materialization happens when joining states?

This code looks dangerous to me. It would permanently mutate whatever generator was captured in `lo`, right? Shouldn't rather mutate the current `generator`? Or is there an invariant that all those generators are identical?

NTillmann replied:
@hermanventer ?

hermanventer replied:
I'm trying to move towards a world where Effects is immutable. This code certainly will not help me get there. I think it should be OK to materialize the leaked objects in the the current generator at this point. Of course, only do so if the appendGenerator flag is true.

Hopefully this will allow the code to become simpler, which is something I'm always in favor of.

sb98052 replied:
I considered several alternatives, and the motivation for capturing the generator at the leak point, creating a LeakedObject type, and modifying the Effects structure was to pass it here, as this is the point at which all of the information needed for materialization comes together. If you materialize into the current generator at this point, you will see the object being built post-join. If you materialize where bindings and properties are joint, you will see disembodied objects missing property bindings, and unexpected deletes because descriptors that `applyEffects` eventually populates are not present at that point. This deletion logic is quite involved and I am glad I avoided colliding with it with the approach that I took.

sb98052 replied:
I have added a comment to that code.

sb98052 commented:
I've addressed the latest batch of review comments.

NTillmann commented on src/realm.js:
```diff
@@ -1485,6 +1490,20 @@ export class Realm {
+    // created for them.
+    if (appendGenerator && leakedObjects !== undefined && leakedObjects.size !== 0) {
+      for (let lo of leakedObjects) {
+        Materialize.materialize(this, lo);
```
It would be good to audit all calls to `applyEffects` to understand the `appendGenerator` logic, and whether it's safe.

sb98052 replied:
See #2295

sb98052 commented:
Abandoning this pull request. The only part left that has not been merged is "making leaking unconditoinal." It is not clear how the implementation in this request factors into the changes coming up in LVA, or if at all. To rethink.
